ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/transactions/TransactionPoolReplacementHandler.java

```diff
@@ -14,35 +14,50 @@
  */
 package org.hyperledger.besu.ethereum.eth.transactions;
 
-import static java.util.Arrays.asList;
-
 import org.hyperledger.besu.ethereum.core.BlockHeader;
 import org.hyperledger.besu.ethereum.eth.transactions.PendingTransactions.TransactionInfo;
+import org.hyperledger.besu.ethereum.mainnet.ProtocolSchedule;
 import org.hyperledger.besu.util.number.Percentage;
 
 import java.util.List;
+import java.util.Map;
+import java.util.function.Function;
+import java.util.stream.Collectors;
 
 import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableMap;
 
 public class TransactionPoolReplacementHandler {
-  private final List<TransactionPoolReplacementRule> rules;
+  private final Map<TransactionPoolReplacementRule, Function<BlockHeader, Boolean>> rules;
+
+  public TransactionPoolReplacementHandler(
+      final ProtocolSchedule protocolSchedule, final Percentage priceBump) {
 
-  public TransactionPoolReplacementHandler(final Percentage priceBump) {
-    this(asList(new TransactionReplacementByPriceRule(priceBump)));
+    this.rules =
+        ImmutableMap.of(
```

Comment:
I was thinking of another way (https://github.com/hyperledger/besu/blob/1772f8b1f89686e674e3ccaf333ae7f6355d1d36/ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/MainnetProtocolSpecs.java#L111) to do it because like that we have behavior logic based on the milestone in two different places in the code

I think we can create a transactionPoolRplacementHandler based on the milestone : 

```java
static ProtocolSpecBuilder londonDefinition(
      final Optional<BigInteger> chainId,
      final OptionalInt configContractSizeLimit,
      final OptionalInt configStackSizeLimit,
      final boolean enableRevertReason,
      final GenesisConfigOptions genesisConfigOptions,
      final boolean quorumCompatibilityMode) {
    ...
    return berlinDefinition(
            chainId,
            configContractSizeLimit,
            configStackSizeLimit,
            enableRevertReason,
            quorumCompatibilityMode)
        .gasCalculator(LondonGasCalculator::new)
        ....
        .transactionPoolReplacementHandler(asList(new TransactionReplacementByGasPriceRule(priceBump),  new TransactionReplacementByFeeMarketRule(priceBump)))
        ....      
        .name(LONDON_FORK_NAME);
  }
```


```java
static ProtocolSpecBuilder frontierDefinition
    ...
    return frontierDefinition(
        ....
        .transactionPoolReplacementHandler(asList(new TransactionReplacementByGasPriceRule(priceBump)))
        ....      
        .name(FRONTIER_FORK_NAME);
  }
```

Reply:
I was reluctant to move the TransactionPoolReplacementRule interface into `core`, but creating a handler interface in there is more palatable.  :+1: